### PR TITLE
Duration fix

### PR DESCRIPTION
**Pull request by JorisCos**
### About this PR
This PR implements `forced_protected_labels` that behaves like `protected_labels` but overrides the soundscape duration to ensure that the foreground are not cut.

Tests also have been implemented to keep 100 %coverage

### CODE REVIEW

File: scaper/core.py [hunk 1]
```diff
@@ -205,6 +205,7 @@ def generate_from_jams(jams_infile,
     # Set synthesis parameters
     if 'sr' in ann.sandbox.scaper: # backwards compatibility
         sc.sr = ann.sandbox.scaper['sr']
+    # sc.forced_protected_labels = ann.sandbox.scaper['forced_protected_labels']
```
Comment (mhu-coder) [inline scaper/core.py:208]: Is this comment necessary?

File: scaper/core.py [hunk 2]
```diff
@@ -1036,8 +1040,12 @@ def __init__(self, duration, fg_path, bg_path, protected_labels=[], random_state
         _populate_label_list(self.fg_path, self.fg_labels)
         _populate_label_list(self.bg_path, self.bg_labels)
 
+        # forced protected labels behave as protected label but they override
+        # the soundscape if they are longer.
+        self.forced_protected_labels = []
+
         # Copy list of protected labels
-        self.protected_labels = protected_labels[:]
+        self.protected_labels = protected_labels[:] + self.forced_protected_labels
```
Comment (mhu-coder) [inline scaper/core.py]: nitpick: the `+` operator creates a new list with the concatenated lists. In other words, I don't think it's necessary to keep the `[:]` after `protected_labels`.

By the way, I noticed that users of the class are expected to set `forced_protected_labels` once the class has been instantiated but wouldn't it make more sense to have `forced_protected_labels` as an argument? If you do add `forced_protected_labels`, please can you avoid using an empty list as the default value (see [example](https://docs.python-guide.org/writing/gotchas/) for an explanation)?

Since `self.protected_labels` is only ever used for lookup as far as I understand, it would make more sense to have a set instead of a list (O(1) lookup vs O(n)). But I understand this would introduce too many changes with respect to the original code.

File: scaper/core.py [hunk 3]
```diff
@@ -1022,6 +1023,9 @@ def __init__(self, duration, fg_path, bg_path, protected_labels=[], random_state
         self.fg_spec = []
         self.bg_spec = []
 
+        # Start with 0 for max_duration
```
Comment (mhu-coder) [inline scaper/core.py]: I am not sure to understand what this comment means. From the code, I understand that `self.ini_duration` is the duration that has been set when a `Scaper` object is instantiated so that this initial duration can be restored later.

Do you think there could be a better way of phrasing this comment?

File: scaper/core.py [hunk 4]
```diff
@@ -1708,6 +1738,7 @@ def _instantiate(self, allow_repeated_label=True,
             fg_labels=self.fg_labels,
             bg_labels=self.bg_labels,
             protected_labels=self.protected_labels,
+            # forced_protected_labels=self.forced_protected_labels,
```
Comment (mhu-coder) [inline scaper/core.py:1741]: Is this line necessary?

File: tests/test_core.py [hunk 5]
```diff
@@ -1301,6 +1301,37 @@ def test_scaper_add_event():
...
+    fg_event = EventSpec(label=('const', 'siren'),
+                         source_file=('choose', []),
+                         source_time=('const', 0),
+                         event_time=('uniform', 0, 9),
+                         event_duration=('truncnorm', 2, 1, 1, 3),
+                         snr=('uniform', 10, 20),
+                         role='foreground',
+                         pitch_shift=('normal', 0, 1),
+                         time_stretch=('uniform', 0.8, 1.2))
+    # Protected labels override the soundscape duration if they are longer
```
Comment (mhu-coder) [inline tests/test_core.py]: Shouldn't it be "Forced protected labels", as opposed to simply "protected labels", which does exist but doesn't change `scaper.duration` (and eventually truncates the event)

**Review by mhu-coder — COMMENTED** [review]
If I understand the PR correctly, the aim is to fix the case where `duration` is smaller than that of a foreground which has been marked as `protected_labels`. Without this fix, the audio of the protected foreground is truncated.

Your proposal is to introduce a `forced_protected_labels` attribute to ensure that such a case results in a soundscape which will contain the complete audio.

I see the following advantage and disadvantage:
* Advantages: the foreground event, which has been declared as `forced_protected` is indeed protected. This would also be the behaviour I would expect for a label that I marked as `protected`
* Disadvantage: the generated soundscape does not have the duration that was specified in `duration`. When specifying a certain duration for a soundscape, I would expected the requirement to be met.

In other words, in the programme, given a protected label, I would first look for the files that meet the duration requirement and only ever use these when generating a soundscape. If there are no such files, an error should be raised.

File: scaper/core.py (hunk 4 above)
Comment (JorisCos) [reply scaper/core.py:1741]: It's not necessary but let it on purpose. I don't add it to the annotation otherwise some test breaks because the key is not in the jams that are already uploaded in test. I will put it as a TODO

File: scaper/core.py (hunk 3 above)
Comment (JorisCos) [reply scaper/core.py]: I forgot to delete this comment line

File: scaper/core.py (hunk 2 above)
Comment (JorisCos) [reply scaper/core.py]: >  But I understand this would introduce too many changes with respect to the original code.

Yep

File: scaper/core.py (hunk 1 above)
Comment (JorisCos) [reply scaper/core.py:208]: It's not necessary but let it on purpose. I don't add it to the annotation otherwise some test breaks because the key is not in the jams that are already uploaded in test. I will put it as a TODO

File: scaper/core.py [hunk 6]
```diff
@@ -1040,12 +1041,15 @@ def __init__(self, duration, fg_path, bg_path, protected_labels=[], random_state
         _populate_label_list(self.fg_path, self.fg_labels)
         _populate_label_list(self.bg_path, self.bg_labels)
 
-        # forced protected labels behave as protected label but they override
+        # forced _protected_labels behave as protected_labels but they override
```
Comment (mhu-coder) [inline scaper/core.py]: `forced_protected_labels` without space?

**Comment by mhu-coder** [comment]
By the way, the test `test_override_duration` fails on `np.allclose`. Is this intended?

File: scaper/core.py (hunk 6 above)
Comment (JorisCos) [reply scaper/core.py]: Thanks

**Comment by JorisCos** [comment]
No it wasn't, It's fixed now

**Comment by JorisCos** [comment]
I think this ready to be merged, let me know if you see anything else
